File: src/main/kotlin/Strings/ReverseString.java
package Strings;

import java.util.Stack;

/**
 * Given an input string s, reverse the order of the words.
 * A word is defined as a sequence of non-space characters. The words in s will be separated by at least one space.
 * Return a string of the words in reverse order concatenated by a single space.
 * Note that s may contain leading or trailing spaces or multiple spaces between two words.
 * The returned string should only have a single space separating the words. Do not include any extra spaces.
 *
 * Example 1:
 * Input: s = "the sky is blue"
 * Output: "blue is sky the"
 * Example 2:
 *
 * Input: s = "  hello world  "
 * Output: "world hello"
 * Explanation: Your reversed string should not contain leading or trailing spaces.
 * Example 3:
 *
 * Input: s = "a good   example"
 * Output: "example good a"
 * Explanation: You need to reduce multiple spaces between two words to a single space in the reversed string.
 *
 *
 * Constraints:
 *
 * 1 <= s.length <= 104
 * s contains English letters (upper-case and lower-case), digits, and spaces ' '.
 * There is at least one word in s.
 *
 * source: leetcode
 */
public class ReverseString {

    public static String getStringInReverseOrder(String str){

        Stack<String> stack = new Stack<>();
        for (String s :
                str.trim().split("\s+")) {
            stack.push(s);
        }

        StringBuilder stringBuilder = new StringBuilder();
        while (!stack.isEmpty()){
            stringBuilder.append(stack.pop());
            stringBuilder.append(" ");
        }
        return stringBuilder.toString().trim();
    }

    public static void main(String[] args) {
        String stringInOrder = "The quick brown fox jumps over the lazy dog";
        System.out.println(getStringInReverseOrder(stringInOrder));
    }
}

File: src/main/kotlin/Strings/ReverseString.kt
package Strings

import Arrays.buyandsell.maxProfit


fun main(args: Array<String>) {
    val string = "The Quick brown fox jump over the lazy dog"
    print(" "+ reverseWords(string))
}
fun reverseWords(s: String): String {
    var start = s.lastIndex
    val dq = ArrayDeque<String>()

    while(start >= 0) {
        var end = start
        while(start >= 0 && s[start] != ' ') start-- // O(N)
        dq.addLast(s.substring(start + 1, end + 1)) // O(N)
        while(start >= 0 && s[start] == ' ') start-- //O(N)
    }

    return dq.joinToString(" ").trim() // O(N)
}